### PR TITLE
[Datasets] [CI] Parallelize Dataset CI tests

### PR DESCRIPTION
**Pull request by scottjlee**
## Why are these changes needed?
Speed up Dataset CI tests by parallelizing test execution, similar to the process recently done for Train in https://github.com/ray-project/ray/pull/29359. In detail:
- Use `parallelism=3` for the Dataset build tests where applicable to parallelize tests.
- Remove coverage for Arrow versions {7, 8, 9, 10} in CI tests (only keep {6, 11, nightly})
- Group tests which require Java into the integrations test group, and remove unnecessary Java installation from the other CI test groups

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by scottjlee** [comment]
By simply using `parallelism=2` for Dataset tests, we see a 50% decrease in test execution time:
- Before: 54m 33s
- After: `max(26m 6s, 27m 23s)` = 27m 23s 

Next step is to break out individual tests in `pipeline.ml.yml` and provide more useful compute size estimation, so that we extract maximum time savings.

**Comment by scottjlee** [comment]
Using `parallelism=3`, the times of the 3 parts are: 22m35s, 17m26s, 15m8s. With a 6 minute setup overhead, we would expect that the entire Dataset CI tests to take about 28 minutes total.

File: .buildkite/pipeline.ml.yml [hunk 1]
```diff
@@ -315,79 +317,50 @@
 - label: "Dataset tests (streaming executor)"
   conditions: ["NO_WHEELS_REQUIRED", "RAY_CI_DATA_AFFECTED"]
   instance_size: medium
+  parallelism: 3
   commands:
     - cleanup() { if [ "${BUILDKITE_PULL_REQUEST}" = "false" ]; then ./ci/build/upload_build_info.sh; fi }; trap cleanup EXIT
     - ./ci/env/install-java.sh
```
Comment (c21) [inline .buildkite/pipeline.ml.yml:323]: can you remove this?

File: .buildkite/pipeline.ml.yml [hunk 2]
```diff
@@ -315,79 +317,50 @@
...
-    - bazel test --config=ci $(./ci/run/bazel_export_options) --build_tests_only --test_tag_filters=ray_data python/ray/air/...
-
-- label: "Dataset tests (Arrow 7)"
+- label: "Dataset tests (Arrow 11)"
   conditions: ["NO_WHEELS_REQUIRED", "RAY_CI_PYTHON_AFFECTED", "RAY_CI_DATA_AFFECTED"]
   instance_size: medium
+  parallelism: 3
   commands:
     - cleanup() { if [ "${BUILDKITE_PULL_REQUEST}" = "false" ]; then ./ci/build/upload_build_info.sh; fi }; trap cleanup EXIT
     - ./ci/env/install-java.sh
```
Comment (c21) [inline .buildkite/pipeline.ml.yml:347]: can you remove this?

File: .buildkite/pipeline.ml.yml [hunk 3]
```diff
@@ -315,79 +317,50 @@
...
+    - ./ci/run/run_bazel_test_with_sharding.sh --config=ci $(./ci/run/bazel_export_options) --build_tests_only --test_tag_filters=-dataset_integration python/ray/data/...
+    - ./ci/run/run_bazel_test_with_sharding.sh --config=ci $(./ci/run/bazel_export_options) --build_tests_only --test_tag_filters=ray_data python/ray/air/...
 
 - label: "Dataset tests (Arrow 6)"
   conditions: ["NO_WHEELS_REQUIRED", "RAY_CI_PYTHON_AFFECTED", "RAY_CI_DATA_AFFECTED"]
   instance_size: medium
+  parallelism: 3
   commands:
     - cleanup() { if [ "${BUILDKITE_PULL_REQUEST}" = "false" ]; then ./ci/build/upload_build_info.sh; fi }; trap cleanup EXIT
     - ./ci/env/install-java.sh
```
Comment (c21) [inline .buildkite/pipeline.ml.yml:359]: can you remove this?

File: .buildkite/pipeline.ml.yml [hunk 4]
```diff
@@ -296,6 +296,8 @@
     - bazel test --config=ci $(./ci/run/bazel_export_options) --build_tests_only python/ray/tests/modin/...
     # Dask tests and examples.
     - bazel test --config=ci $(./ci/run/bazel_export_options) --build_tests_only --test_tag_filters=-client python/ray/util/dask/...
+    # RayDP tests and examples.
+    - bazel test --config=ci $(./ci/run/bazel_export_options) --build_tests_only --test_tag_filters=-client python/ray/data/tests/test_raydp_dataset.py
```
Comment (c21) [inline .buildkite/pipeline.ml.yml:300]: we should also add `./ci/env/install-java.sh`.

**Comment by ollie-iterators** [comment]
The test_task_events failure can be fixed my merging in recent changes to the branch

**Comment by c21** [comment]
hmm where is RayDP test now?

**Comment by scottjlee** [comment]
> hmm where is RayDP test now?

TLDR; they are included in the `:potable_water: Dataset datasource integration tests (Python 3.7)` test now.

I [configured the RayDP tests ](https://github.com/ray-project/ray/pull/33443/files#diff-b5e296091b106d33c72a211245511d79e74808f508df84ccbaee266f4001ce74R374-R380) to have the tag `dataset_integration`, which is used in the `bazel test` filters to (a) exclude in runs which we do not install Java; and (b) include only these tests in the CI test item which installs Java (i.e. [:potable_water: Dataset datasource integration tests (Python 3.7)](https://github.com/ray-project/ray/blob/96945b0b16676d4723f4b395a5d3ca930cd2e880/.buildkite/pipeline.ml.yml#L432)). So

**Comment by ollie-iterators** [comment]
rl_offline_example should be fixed when https://github.com/ray-project/ray/pull/33750 gets merged

**Review by c21 — APPROVED** [review]
LG, thanks @scottjlee! Great speedup.

**Review by clarkzinzow — APPROVED** [review]
Awesome work @scottjlee!

In this PR, we're informally dropping support for Arrow 7, 8, 9, and 10 by dropping their coverage in CI. In a future PR, we might want to make that explicit via the:
- pyarrow version that's installed with Ray: https://github.com/ray-project/ray/blob/c7143398c9e2fe42f6f4e56c97cc41050028e1c8/python/setup.py#L236-L241
- the runtime check that the installed pyarrow is supported: https://github.com/ray-project/ray/blob/c7143398c9e2fe42f6f4e56c97cc41050028e1c8/python/ray/data/_internal/util.py#L24

IMO, this can definitely be deferred to a future PR, but we should probably figure out what we want to advertise as supported Arrow versions for Ray 2.5.

**Comment by c21** [comment]
Can we undo all whitespace removal change?

**Comment by scottjlee** [comment]
> Can we undo all whitespace removal change?

Done! Sorry not sure what happened there... those changes occurred as a result of running `format.sh`, but now I ran them again and it didn't make those changes this time

**Comment by c21** [comment]
@zhe-thoughts - this is a Data CI test change only, to speed up CI test. No actual non-test code change.

**Review by zhe-thoughts — APPROVED** [review]
